crates/codegen/spec/src/snippets.rs

```diff
@@ -51,21 +52,42 @@ impl<'context> Snippets<'context> {
         let production_name = &production.name();
         let (section, topic) = self.grammar.locate_production(production_name);
 
-        let file_name = match production.versions() {
-            None => "unversioned".to_owned(),
-            Some(versions) => {
-                Production::get_exact_version(versions.into_iter(), &version)?.to_string()
-            }
+        let file_name = match production.as_ref() {
+            Production::Scanner { version_map, .. } => match version_map {
+                VersionMap::Unversioned(_) => Some("unversioned".to_owned()),
+                VersionMap::Versioned(versions) => versions
+                    .keys()
+                    .rev()
+                    .find(|v| *v <= version)
+                    .and_then(|v| versions.get(v).unwrap().as_ref().map(|_| v.to_string())),
+            },
+            Production::TriviaParser { version_map, .. }
+            | Production::Parser { version_map, .. } => match version_map {
```

Comment:
Instead of having each consumer of `VersionMap` have that unwrapping logic (currently both `schema` and `spec`), I wonder if it can be encapsulated/shared, like in the former `get_exact_version()`.

Reply:
Possibly, but code refactoring like that I would prefer a separate PR for a big cleanup. I have seen a number of things I would like to do.